pkg/datastore/phpapi/phpapi.go

```diff
@@ -188,10 +188,14 @@ func (p *PHPAPI) SetEndpoint(local types.SubmarinerEndpoint) error {
 	}
 	formVal := url.Values{}
 	formVal.Set("action", "reconcile")
-	formVal.Add("endpoint",string(marshaledEndpoint))
+	formVal.Add("endpoint", string(marshaledEndpoint))
 	formedURL := fmt.Sprintf("%s://%s/endpoints.php?identifier=%s&cluster_id=%s", p.Proto, p.Server, p.APIToken, local.Spec.ClusterID)
 	klog.V(8).Infof("Formed URL was %s", formedURL)
 	poster, err := http.PostForm(formedURL, formVal)
+	if err != nil {
+		klog.Errorf("Error while setting endpoint %v", err)
```

Comment:
Instead of logging the error here, how about returning a formatted error with the added context to the caller instead of just returning err, ie

  return fmt.Errorf("Error while setting endpoint %v", err)

 This would avoid duplicate logging, assuming the caller also logs it, as it probably should and not assume the implementation will log it. However I can see it the other way as well, ie should the implementation assume the caller will appropriately handle and log the returned error. I kind of prefer the former. Either way, we should be consistent - line 87 just returns the err without logging any additional context.

That said, cleaning all that up should probably be a separate patch.

Reply:
I seem to remember having this discussion on ODL too ;-). I tend to favour a “log, or throw, not both” approach (adding useful information to errors as they go up the stack), which matches your preference here; however I also think it’s best to keep things consistent, so fix up the body handling in this patch, and then make logging consistent everywhere later on...

Reply:
I agree with @skitt on this